contracts/src/token/erc20/extensions/flash_mint.rs

```diff
@@ -0,0 +1,448 @@
+            #[allow(missing_docs)]
+            function onFlashLoan(
+                address initiator,
+                address token,
+                uint256 amount,
+                uint256 fee,
+                bytes calldata data
+            ) external returns (bytes32);
+        }
+    }
+}
+
+/// State of the [`Erc20FlashMint`] Contract.
+#[storage]
+pub struct Erc20FlashMint {
+    /// Fee applied when doing flash loans.
+    pub flash_fee_amount: StorageU256,
+    /// Receiver address of the flash fee.
+    pub flash_fee_receiver_address: StorageAddress,
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for Erc20FlashMint {}
+
+/// Interface of the ERC-3156 Flash Lender, as defined in [ERC-3156].
+///
+/// [ERC-3156]: https://eips.ethereum.org/EIPS/eip-3156
+pub trait IErc3156FlashLender {
+    /// The error type associated to this trait implementation.
+    type Error: Into<alloc::vec::Vec<u8>>;
+
+    /// Returns the maximum amount of tokens available for loan.
+    ///
+    /// NOTE: This function does not consider any form of supply cap, so in case
+    /// it's used in a token with a cap like
+    /// [`crate::token::erc20::extensions::Capped`], make sure to override this
+    /// function to integrate the cap instead of [`U256::MAX`].
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `token` - The address of the token that is requested.
+    /// * `erc20` - Read access to an [`Erc20`] contract.
+    ///
+    /// # Examples
+    ///
+    /// In order to have [`IErc3156FlashLender::max_flash_loan`] exposed in ABI,
+    /// you need to do this manually.
+    ///
+    /// ```rust,ignore
+    /// fn max_flash_loan(&self, token: Address) -> U256 {
+    ///     self.erc20_flash_mint.max_flash_loan(token, &self.erc20)
+    /// }
+    fn max_flash_loan(&self, token: Address, erc20: &Erc20) -> U256;
+
+    /// Returns the fee applied when doing flash loans.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `token` - The token to be flash loaned.
+    /// * `amount` - The amount of tokens to be loaned.
+    ///
+    /// # Errors
+    ///
+    /// If the token is not supported, then the error
+    /// [`Error::UnsupportedToken`] is returned.
+    ///
+    /// # Examples
+    ///
+    /// In order to have [`IErc3156FlashLender::flash_fee`] exposed in ABI, you
+    /// need to do this manually.
+    ///
+    /// ```rust,ignore
+    /// fn flash_fee(&self, token: Address, amount: U256) -> Result<U256, Vec<u8>> {
+    ///     Ok(self.erc20_flash_mint.flash_fee(token, amount)?)
+    /// }
+    fn flash_fee(
+        &self,
+        token: Address,
+        amount: U256,
```

Comment:
It is not clear which terminology to use here:
- ERC20 uses `value` (see [transfer](https://eips.ethereum.org/EIPS/eip-20#transfer) def)
- ERC3156 uses `amount` (see [EIP])

The Solidity team had the [same discussion](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3864). They opted for keeping things consistent with ERC20. Our implementation does not; instead it is consistent with the [EIP].

Wdyt?

[EIP]: https://eips.ethereum.org/EIPS/eip-3156#lender-specification

Reply:
Yea, I mentioned that above. Should we align with the Solidity team or with official EIP, wdyt?

Reply:
Let's align with the Solidity team.